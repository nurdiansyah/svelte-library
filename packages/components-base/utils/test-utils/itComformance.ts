import type { RenderResult } from "@testing-library/svelte";
import type { SvelteComponentDev } from "svelte/internal";
import { cleanup, render } from "@testing-library/svelte";

export type BaseProps = {
  "data-testid"?: string;
};

export type Options<P = Record<string, any>> = {
  props: P;
  container?: HTMLElement;
  parentProps?: Record<string, any>;
  after?: () => void;
  only?: string[];
  skip?: string[];
  target?: HTMLElement;
  className?: string;
};

function testMergeClassName(_render: () => RenderResult, { props, className }: Options) {
  it("cek merge class", () => {
    let el: undefined | Element;
    const { getByTestId, container } = _render();
    if (props["data-testid"]) {
      el = getByTestId(props["data-testid"]);
    } else {
      el = container?.firstElementChild || undefined;
    }
    className && expect(el).toHaveClass(className);
    props["class"] && expect(el).toHaveClass(props["class"]);
  });
}

function testRendererComponent(_render: () => RenderResult, { props }: Options) {
  it("cek render komponen tanpa error dan snapshot", () => {
    let el: undefined | HTMLElement;
    const { getByTestId, container } = _render();
    if (props["data-testid"]) {
      expect(() => (el = getByTestId(props["data-testid"]))).not.toThrow();
      expect(el).toBeInTheDocument();
    }
    expect(container).toMatchSnapshot();
  });
}

const fullSuite: Record<string, (_render: () => RenderResult, options: Options) => void> = {
  renderComponent: testRendererComponent,
  mergeClassName: testMergeClassName
};

export function itConformance<P = Record<string, any>>(component: typeof SvelteComponentDev, options: Options<P>) {
  const { after: runAfterHook = () => {}, only = Object.keys(fullSuite), skip = [], props = {}, container } = options;
  const renderComponent = (_props?: P): RenderResult => {
    const __props = { ...props, ..._props };

    return render(component, __props, { container });
  };
  afterEach(() => {
    cleanup();
    runAfterHook && runAfterHook();
  });

  Object.keys(fullSuite)
    .filter((testKey) => only.indexOf(testKey) !== -1 && skip.indexOf(testKey) === -1)
    .forEach((testKey) => {
      const test = fullSuite[testKey];
      test(renderComponent, options);
    });
  return renderComponent;
}
