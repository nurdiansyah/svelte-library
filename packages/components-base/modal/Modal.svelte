<script context="module">
  // TODO fade option
  let openCount = 0;
</script>

<script lang="ts">
  import clsx from "clsx";
  import { browserEvent } from ".../utils";
  import { onDestroy, onMount, afterUpdate } from "svelte";
  import { fade as fadeTransition } from "svelte/transition";
  import { conditionallyUpdateScrollbar, getOriginalBodyPadding, setScrollbarWidth } from ".../utils";

  function noop() {}

  let className: string | null = null,
    classes: string;
  export { className as class };
  export let isOpen: boolean = false;
  export let autoFocus: boolean = true;
  export let centered: boolean = false;
  export let backdropDuration: number = 0;
  export let scrollable: boolean = false;
  export let size: "sm" | "lg";
  export let toggle: Function;
  export let labelledBy: string | null = null;
  export let backdrop: boolean = true;
  export let onEnter = undefined;
  export let onExit = undefined;
  export let onOpened = noop;
  export let onClosed = noop;
  export let wrapClassName: string | null = null;
  export let modalClassName: string | null = null;
  export let backdropClassName: string | null = null;
  export let contentClassName: string | null = null;
  export let fade: boolean = true;
  export let zIndex: number = 1050;
  export let unmountOnClose: boolean = true;
  export let returnFocusAfterClose: boolean = true;
  export let transitionType = fadeTransition;
  export let transitionOptions = {};

  let hasOpened = false;
  let _isMounted = false;
  let _triggeringElement;
  let _originalBodyPadding;
  let _lastIsOpen = isOpen;
  let _lastHasOpened = hasOpened;
  let _dialog;
  let _mouseDownElement;
  let _removeEscListener;

  onMount(() => {
    if (isOpen) {
      init();
      hasOpened = true;
    }

    if (typeof onEnter === "function") {
      onEnter();
    }

    if (hasOpened && autoFocus) {
      setFocus();
    }
  });

  onDestroy(() => {
    if (typeof onExit === "function") {
      onExit();
    }

    destroy();
    if (hasOpened) {
      close();
    }
  });

  afterUpdate(() => {
    if (isOpen && !_lastIsOpen) {
      init();
      hasOpened = true;
    }

    if (autoFocus && hasOpened && !_lastHasOpened) {
      setFocus();
    }

    _lastIsOpen = isOpen;
    _lastHasOpened = hasOpened;
  });

  function setFocus() {
    if (_dialog && _dialog.parentNode && typeof _dialog.parentNode.focus === "function") {
      _dialog.parentNode.focus();
    }
  }

  function init() {
    try {
      _triggeringElement = document.activeElement;
    } catch (err) {
      _triggeringElement = null;
    }

    _originalBodyPadding = getOriginalBodyPadding();
    conditionallyUpdateScrollbar();
    if (openCount === 0) {
      document.body.className = clsx(document.body.className, "modal-open");
    }

    ++openCount;
    _isMounted = true;
  }

  function manageFocusAfterClose() {
    if (_triggeringElement) {
      if (typeof _triggeringElement.focus === "function" && returnFocusAfterClose) {
        _triggeringElement.focus();
      }

      _triggeringElement = null;
    }
  }

  function destroy() {
    manageFocusAfterClose();
  }

  function close() {
    if (openCount <= 1) {
      const modalOpenClassName = "modal-open";
      const modalOpenClassNameRegex = new RegExp(`(^| )${modalOpenClassName}( |$)`);
      document.body.className = document.body.className.replace(modalOpenClassNameRegex, " ").trim();
    }

    manageFocusAfterClose();
    openCount = Math.max(0, openCount - 1);

    setScrollbarWidth(_originalBodyPadding);
  }

  function handleBackdropClick(e) {
    if (e.target === _mouseDownElement) {
      e.stopPropagation();
      if (!isOpen || !backdrop) {
        return;
      }

      const backdropElem = _dialog ? _dialog.parentNode : null;
      if (backdropElem && e.target === backdropElem && toggle) {
        toggle(e);
      }
    }
  }

  function onModalOpened() {
    _removeEscListener = browserEvent(document, "keydown", (event) => {
      if (event.key && event.key === "Escape") {
        toggle(event);
      }
    });

    onOpened();
  }

  function onModalClosed() {
    onClosed();

    if (_removeEscListener) {
      _removeEscListener();
    }

    if (unmountOnClose) {
      destroy();
    }
    close();
    if (_isMounted) {
      hasOpened = false;
    }
    _isMounted = false;
  }

  function handleBackdropMouseDown(e) {
    _mouseDownElement = e.target;
  }

  const dialogBaseClass = "modal-dialog";

  $: classes = clsx(dialogBaseClass, className, {
    [`modal-${size}`]: size,
    [`${dialogBaseClass}-centered`]: centered,
    [`${dialogBaseClass}-scrollable`]: scrollable
  });
</script>

{#if _isMounted}
  <div {...$$restProps} class="{wrapClassName}" tabindex="-1" style="position: relative; z-index: {zIndex}">
    {#if isOpen}
      <div
        transition:transitionType="{transitionOptions}"
        ariaLabelledby="{labelledBy}"
        class="{classNames('modal', 'show', modalClassName)}"
        role="dialog"
        style="display: block;"
        on:introend="{onModalOpened}"
        on:outroend="{onModalClosed}"
        on:click="{handleBackdropClick}"
        on:mousedown="{handleBackdropMouseDown}">
        <div class="{classes}" role="document" bind:this="{_dialog}">
          <div class="{classNames('modal-content', contentClassName)}">
            <slot name="external" />
            <slot />
          </div>
        </div>
      </div>
      <div
        transition:fadeTransition="{{ duration: fade && backdropDuration }}"
        class="{classNames('modal-backdrop', 'show', backdropClassName)}"></div>
    {/if}
  </div>
{/if}

<style lang="scss" global>
  @use "./modal.scss";
</style>
