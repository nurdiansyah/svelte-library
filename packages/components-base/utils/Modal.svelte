<script lang="ts">
  import type { SvelteComponent } from "svelte/internal";

  import { clsx } from "./clsx";
  import { onMount, onDestroy } from "svelte";
  import { fade as fadeTransition } from "svelte/transition";
  import TrapFocus from "./modal/TrapFocus.svelte";
  import { ownerDocument } from "./dom";
  import Portal from "./Portal.svelte";
  import { getModalManager, ModalManager } from "./modal/ModalManager";
  import SimpleBackdrop from "./modal/SimpleBackdrop.svelte";

  let { class: className } = $$props;
  let classes: string;
  let exited: boolean = false;
  let BackdropComponent: typeof SvelteComponent | undefined;

  export let backdropComponent: typeof SvelteComponent = SimpleBackdrop;
  export let backdropProps: Record<string, any> = {};
  export let closeAfterTransition: boolean = false;
  export let container: HTMLElement | undefined = undefined;
  export let disableAutoFocus: boolean = false;
  export let disableBackdropClick: boolean = false;
  export let disableEnforceFocus: boolean = false;
  export let disableEscapeKeyDown: boolean = false;
  export let disablePortal: boolean = false;
  export let disableRestoreFocus: boolean = false;
  export let disableScrollLock: boolean = false;
  export let hideBackdrop: boolean = false;
  export let labelledBy: string | undefined = undefined;
  export let manager: ModalManager | undefined = undefined;
  export let node: HTMLElement | undefined = undefined;
  export let onBackdropClick: null | ((e?: Event) => void) = null;
  export let onEnter: null | ((e?: Event) => void) = null;
  export let onEntering: null | ((e?: Event) => void) = null;
  export let onEntered: null | ((e?: Event) => void) = null;
  export let onExit: null | ((e?: Event) => void) = null;
  export let onExiting: null | ((e?: Event) => void) = null;
  export let onExited: null | ((e?: Event) => void) = null;
  export let onClose: null | ((e?: Event, info?: any) => void) = null;
  export let onEscapeKeyDown: null | ((e: Event) => void) = null;
  export let onOpen: null | ((e?: Event) => void) = null;
  export let open: boolean = false;
  export let transitionDuration: number = 300;

  const isTopModal = () => manager?.isTopModal(node) || false;

  function handleMount() {
    if (!manager) {
      manager = getModalManager();
    }
    manager.mount(node, { disableScrollLock });
    if (node) node.scrollTop = 0;
  }

  function handleOpen() {
    if (!manager || !node || !container) return;
    manager.add(node, container);
    if (onOpen) {
      onOpen();
    }
    if (exited) {
      handleMount();
    }
  }

  function handleClose() {
    const index = (manager && node && manager.remove(node)) || -1;
    if (index > -1 && onClose) {
      onClose();
    }
  }

  function handleBackdropClick(e: Event) {
    if (e.target !== e.currentTarget) {
      return;
    }
    if (onBackdropClick) {
      onBackdropClick(e);
    }
    if (!disableBackdropClick && onClose) {
      onClose(e, "backdrop-click");
    }
  }

  function handleKeyDown(e: KeyboardEvent) {
    // The handler doesn't take event.defaultPrevented into account:
    //
    // event.preventDefault() is meant to stop default behaviours like
    // clicking a checkbox to check it, hitting a button to submit a form,
    // and hitting left arrow to move the cursor in a text input etc.
    // Only special HTML elements have these default behaviors.
    if (e.key !== "Escape" || !isTopModal()) {
      return;
    }

    // Swallow the event, in case someone is listening for the escape key on the body.
    e.stopPropagation();

    if (onEscapeKeyDown) {
      onEscapeKeyDown(e);
    }

    if (!disableEscapeKeyDown && onClose) {
      onClose(e);
    }
  }

  function handleEntering() {
    onEntering && onEntering();
  }
  function handleEntered() {
    onEntered && onEntered();
  }
  function handleExiting() {
    onExiting && onExiting();
  }
  function handleExited() {
    closeAfterTransition && handleClose();
    onExited && onExited();
    exited = true;
  }

  onMount(async () => {
    if (!container) container = ownerDocument(node).body;
    onEnter && onEnter();
  });

  onDestroy(() => {
    onExit && onExit();
  });

  $: {
    if (open) {
      handleOpen();
    } else if (!closeAfterTransition) {
      handleClose();
    }
  }
  $: classes = clsx(className, "dbx-modal", open && "-open");

  $: {
    BackdropComponent = hideBackdrop ? undefined : backdropComponent;
  }
</script>

{#if container}
  <Portal {container} disable={disablePortal}>
    <div
      {...$$restProps}
      class={classes}
      aria-labelledby={labelledBy}
      bind:this={node}
      transition:fadeTransition|local={{ duration: transitionDuration }}
      on:outroend={handleExited}
      on:outrostart={handleExiting}
      on:introstart={handleEntering}
      on:introend={handleEntered}
      on:keydown={handleKeyDown}
      role="presentation">
      <svelte:component this={BackdropComponent} {...backdropProps} on:click={handleBackdropClick} />
      <TrapFocus
        {disableAutoFocus}
        {disableEnforceFocus}
        {disableRestoreFocus}
        enable={isTopModal()}
        {open}
        getDoc={() => ownerDocument(node)}>
        <slot />
      </TrapFocus>
    </div>
  </Portal>
{/if}

<style src="./styles/modal.scss" global>
</style>
