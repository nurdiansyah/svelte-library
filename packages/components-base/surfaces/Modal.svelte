<script context="module">
  // TODO fade option
  let openCount = 0;
</script>

<script lang="ts">
  import clsx from "clsx";
  import { onDestroy, onMount, afterUpdate } from "svelte";
  import { fade as fadeTransition } from "svelte/transition";
  import { domUtils } from "../utils";
  import Header from "./modal/Header.svelte";
  import Footer from "./modal/Footer.svelte";

  const { getOriginalBodyPadding, setScrollbarWidth } = domUtils;

  let className: string | null = null,
    classes: string;
  export { className as class };
  export let isOpen: boolean = false;
  export let autoFocus: boolean = true;
  export let centered: boolean = false;
  export let backdropDuration: number = 0;
  export let scrollable: boolean = false;
  export let size: "sm" | "lg";
  export let toggle: Function;
  export let labelledBy: string | null = null;
  export let backdrop: boolean = true;
  export let onEnter: null | ((e: Event) => void) = null;
  export let onExit: null | ((e: Event) => void) = null;
  export let onOpened: null | ((e: Event) => void) = null;
  export let onClosed: null | ((e: Event) => void) = null;
  export let wrapClassName: string | null = null;
  export let modalClassName: string | null = null;
  export let backdropClassName: string | null = null;
  export let contentClassName: string | null = null;
  export let fade: boolean = true;
  export let zIndex: number = 1050;
  export let unmountOnClose: boolean = true;
  export let returnFocusAfterClose: boolean = true;
  export let transitionType = fadeTransition;
  export let transitionOptions = {};

  let hasOpened = false;
  let _isMounted = false;
  let _triggeringElement;
  let _originalBodyPadding;
  let _lastIsOpen = isOpen;
  let _lastHasOpened = hasOpened;
  let _dialog;
  let _mouseDownElement;
  let _removeEscListener;

  onMount(() => {
    if (isOpen) {
      init();
      hasOpened = true;
    }

    if (typeof onEnter === "function") {
      onEnter();
    }

    if (hasOpened && autoFocus) {
      setFocus();
    }
  });

  onDestroy(() => {
    if (typeof onExit === "function") {
      onExit();
    }

    destroy();
    if (hasOpened) {
      close();
    }
  });

  afterUpdate(() => {
    if (isOpen && !_lastIsOpen) {
      init();
      hasOpened = true;
    }

    if (autoFocus && hasOpened && !_lastHasOpened) {
      setFocus();
    }

    _lastIsOpen = isOpen;
    _lastHasOpened = hasOpened;
  });

  function setFocus() {
    if (_dialog && _dialog.parentNode && typeof _dialog.parentNode.focus === "function") {
      _dialog.parentNode.focus();
    }
  }

  function init() {
    try {
      _triggeringElement = document.activeElement;
    } catch (err) {
      _triggeringElement = null;
    }

    _originalBodyPadding = getOriginalBodyPadding();
    if (openCount === 0) {
      document.body.className = clsx(document.body.className, "modal-open");
    }

    ++openCount;
    _isMounted = true;
  }

  function manageFocusAfterClose() {
    if (_triggeringElement) {
      if (typeof _triggeringElement.focus === "function" && returnFocusAfterClose) {
        _triggeringElement.focus();
      }

      _triggeringElement = null;
    }
  }

  function destroy() {
    manageFocusAfterClose();
  }

  function close() {
    if (openCount <= 1) {
      const modalOpenClassName = "modal-open";
      const modalOpenClassNameRegex = new RegExp(`(^| )${modalOpenClassName}( |$)`);
      document.body.className = document.body.className.replace(modalOpenClassNameRegex, " ").trim();
    }

    manageFocusAfterClose();
    openCount = Math.max(0, openCount - 1);

    setScrollbarWidth(_originalBodyPadding);
  }

  function handleBackdropClick(e) {
    if (e.target === _mouseDownElement) {
      e.stopPropagation();
      if (!isOpen || !backdrop) {
        return;
      }

      const backdropElem = _dialog ? _dialog.parentNode : null;
      if (backdropElem && e.target === backdropElem && toggle) {
        toggle(e);
      }
    }
  }

  function onModalOpened() {
    _removeEscListener = browserEvent(document, "keydown", (event) => {
      if (event.key && event.key === "Escape") {
        toggle(event);
      }
    });

    onOpened();
  }

  function onModalClosed() {
    onClosed();

    if (_removeEscListener) {
      _removeEscListener();
    }

    if (unmountOnClose) {
      destroy();
    }
    close();
    if (_isMounted) {
      hasOpened = false;
    }
    _isMounted = false;
  }

  function handleBackdropMouseDown(e) {
    _mouseDownElement = e.target;
  }

  $: classes = clsx(
    className,
    "modal-dialog",
    size && `-${size}`,
    centered && "-centered",
    scrollable && "-scrollable"
  );
</script>

{#if _isMounted}
  <div {...$$restProps} class={wrapClassName} tabindex="-1" style="position: relative; z-index: {zIndex}">
    {#if isOpen}
      <div
        transition:transitionType={transitionOptions}
        ariaLabelledby={labelledBy}
        class={clsx('dbx-modal', '-show', modalClassName)}
        role="dialog"
        style="display: block;"
        on:introend={onModalOpened}
        on:outroend={onModalClosed}
        on:click={handleBackdropClick}
        on:mousedown={handleBackdropMouseDown}>
        <div class={classes} role="document" bind:this={_dialog}>
          <div class={clsx('modal-content', contentClassName)}>
            <slot name="header" class="header" />
            {#if $$slots.default}
              <div class="content">
                <slot />
              </div>
            {/if}
            {#if $$slots.header}
              <Footer {...footerProps}>
                <slot name="footer" />
              </Footer>
            {/if}
          </div>
        </div>
      </div>
      <div
        transition:fadeTransition={{ duration: fade && backdropDuration }}
        class={classNames('modal-backdrop', 'show', backdropClassName)} />
    {/if}
  </div>
{/if}

<style src="./styles/modal.scss" global>
</style>
