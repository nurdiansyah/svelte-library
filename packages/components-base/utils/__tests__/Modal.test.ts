import Modal from "./mocks/ModalMock.svelte";
import { render } from "@testing-library/svelte";
// import { logDOM } from "@testing-library/dom";
import { BaseProps, itConformance } from "../test-utils";

describe("Modal", () => {
  type Props = BaseProps & { open?: true };
  const testId = "modal";

  itConformance<Props>(Modal, {
    props: {
      "data-testid": testId
    }
  });
  it("check: props => open", () => {
    const _render = (props: Props) => render(Modal, { props });
    const { getByTestId } = _render({ open: true });
    expect(() => getByTestId(testId)).not.toThrow();
    // logDOM(container);
  });
});
