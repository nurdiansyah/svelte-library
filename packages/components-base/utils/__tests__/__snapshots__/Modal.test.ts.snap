// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal cek render komponen tanpa error dan snapshot 1`] = `
<body>
  <div>
    <div
      style="display: none;"
    />
    
    
  </div>
  <div>
    <div
      class="dbx-modal -open"
      data-testid="modal"
      role="presentation"
    >
       
      <div>
        <div
          data-testid="sentinel-start"
          tabindex="0"
        />
         
         
        <div
          data-testid="sentinel-end"
          tabindex="0"
        />
      </div>
    </div>
  </div>
</body>
`;
